subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/BuildCacheClasspathEntrySnapshotCache.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2014 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.classpath;
+
+import org.gradle.caching.BuildCacheKey;
+import org.gradle.caching.internal.controller.BuildCacheController;
+import org.gradle.caching.internal.controller.BuildCacheLoadCommand;
+import org.gradle.caching.internal.controller.BuildCacheStoreCommand;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.hash.Hashing;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder;
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder;
+import org.gradle.util.GradleVersion;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Stores classpath entry snapshots in the build cache and loads them from there when they can't be found in the given delegate.
+ */
+public class BuildCacheClasspathEntrySnapshotCache implements ClasspathEntrySnapshotCache {
```

Comment:
@wolfs I'd like your feedback on this class - This pattern could be generalized into a build cache decorator for any kind of expensive data. For example, we could also use this pattern to store the Classpath/CompileClasspath snapshots for external dependencies, so we don't have to recompute them every time (e.g. on ephemeral CI nodes this can cost a lot of time).

Reply:
You should probably use the new `ExecutionEngine` instead. The idea would be to create a `UnitOfWork` to perform the class analysis. The engine should handle storing in the cache for you, so you don't need to implement your own cache keys for example.